### PR TITLE
feat: receivable_msat to listpeers

### LINKED ISSUE #3555: Feature request: receivable_msat
### Add `receivable_msat` to `listpeers`  channels output.
In order to know how much is currently able to receive on a channel (with one transaction) we may want to have `receivable_msat` /  `receivable_msatoshi` in `listpeers` output similar to `spendable_msat`.

A remote may still reject to send that much through a channel for other reasons, but `receivable_msat` should reflect the technical maximum: `total - reserves - 1x funders on-chain HTLC fees at current rate.`

Among others this will allow to finish drain plugin, specifically the `fill` command, in a clean way: https://github.com/lightningd/plugins/pull/71

### PR DESCRIPTION
**Pull request by m-schmoock**
This adds the `receivable_msat` amount to the `lightning-cli listpeers` output. The logic is similar to `spendable_msat` and applies for the maximum possible amount in one HTLC in that situation.

Adds code, test and manpage - Closes #3555

### CODE REVIEW

File: lightningd/peer_control.c [hunk 1]
```diff
@@ -573,13 +573,30 @@ static void subtract_offered_htlcs(const struct channel *channel,
 	}
 }
 
+static void subtract_received_htlcs(const struct channel *channel,
+				   struct amount_msat *amount)
```
Comment (niftynei) [inline lightningd/peer_control.c]: nit: this should align with the above 

```
(const
 struct amount_msat
```

File: tests/test_pay.py [hunk 2]
```diff
@@ -2245,6 +2245,59 @@ def test_channel_spendable(node_factory, bitcoind):
...
+    assert l2.rpc.listpeers()['peers'][0]['channels'][0]['receivable_msat'] == Millisatoshi(0)
+    l1.rpc.waitsendpay(payment_hash, TIMEOUT)
+
+    # Make sure l2 thinks it's all over.
+    wait_for(lambda: len(l2.rpc.listpeers()['peers'][0]['channels'][0]['htlcs']) == 0)
+    # Now, reverse should work similarly.
+    payment_hash = l1.rpc.invoice('any', 'inv', 'for testing')['payment_hash']
+    amount = l1.rpc.listpeers()['peers'][0]['channels'][0]['receivable_msat']
+
+    # Turns out we this won't route, as it's over max - reserve:
```
Comment (niftynei) [inline tests/test_pay.py]:
```suggestion
    # Turns out we won't route this, as it's over max - reserve:
```

File: lightningd/peer_control.c [hunk 3]
```diff
@@ -692,33 +709,60 @@ static void json_add_channel(struct lightningd *ld,
...
+	if (!amount_msat_sub_sat(&receivable,
+				 their_msat,
+				 channel->our_config.channel_reserve))
+		receivable = AMOUNT_MSAT(0);
+	/* Take away any currently-offered HTLCs. */
+	subtract_received_htlcs(channel, &receivable);
+	/* If they're funder, subtract txfees they'll need to spend this */
+	if (channel->funder == REMOTE) {
+		if (!amount_msat_sub_sat(&receivable, receivable,
+					 commit_txfee(channel, receivable)))
```
Comment (m-schmoock) [inline lightningd/peer_control.c]: Hm, I'm sure we need to modify `commit_txfee(channel, spendable)` or add an analogue function that calculates this for the remote side...Will work something out...  The unit tests didn't find this, since fee configs are the same for every testtnode...

Comment (m-schmoock) [reply lightningd/peer_control.c]: Im reviewing myself here, please wait until I addressed this one :D

Comment (m-schmoock) [reply lightningd/peer_control.c]: addressed by 99948c9cf

File: tests/test_pay.py [hunk 4]
```diff
@@ -2256,54 +2309,55 @@ def test_channel_spendable_large(node_factory, bitcoind):
...
+    l1.force_feerates(22500)
     l2.pay(l1, total // 2)
 
+    # reset fees and send all back again
+    l1.force_feerates(15000)
+    l1.drain(l2)
+
+    # But if feerate increase just a little more, l2 should not be able to send
+    # non-fust HTLC to l1
+    l1.force_feerates(22502)  # TODO: Why does 22501 fail? off by one in C code?
```
Comment (m-schmoock) [inline tests/test_pay.py]: ideas?

**Comment by m-schmoock** [comment]
From my side, I'm more or less okay with this PR, aside I dont like the way I needed duplicated code (i.e. commit_txfee_spend/recv) with all just the `REMOTE` / `LOCAL`   or  `our_config` / `their_config`  changes. But any approach to unify the blocks lead to a iffy mess.

Also I think there might be an off by one in current (master) implementation, see https://github.com/ElementsProject/lightning/pull/3572/files/7d1f8416e1af8bf9579914db46da6ce5b5cb394a#diff-e6fd9e4a383e2ab7819a6292c1edf70bR2357 for my remark in the testcase.

File: lightningd/peer_control.c [hunk 5]
```diff
@@ -692,33 +709,60 @@ static void json_add_channel(struct lightningd *ld,
 				   channel->channel_info.their_config.channel_reserve,
 				   "our_channel_reserve_satoshis",
 				   "our_reserve_msat");
-	/* Compute how much we can send via this channel. */
+
+
```
Comment (cdecker) [inline lightningd/peer_control.c]: Nit: Single empty line is sufficient.

File: lightningd/peer_control.c [hunk 6]
```diff
@@ -692,33 +709,60 @@ static void json_add_channel(struct lightningd *ld,
...
-
 	/* We can't offer an HTLC over the max payment threshold either. */
 	if (amount_msat_greater(spendable, chainparams->max_payment))
 		spendable = chainparams->max_payment;
-
+	/* append spendable to JSON output */
 	json_add_amount_msat_compat(response, spendable,
 				    "spendable_msatoshi", "spendable_msat");
+
+
```
Comment (cdecker) [inline lightningd/peer_control.c]: not: Too much whitespace. You also remove the above whitespace which I would have liked to stay.

File: lightningd/peer_control.c [hunk 7]
```diff
@@ -692,33 +709,60 @@ static void json_add_channel(struct lightningd *ld,
...
+		receivable = AMOUNT_MSAT(0);
+	/* Take away any currently-offered HTLCs. */
+	subtract_received_htlcs(channel, &receivable);
+	/* If they're funder, subtract txfees they'll need to spend this */
+	if (channel->funder == REMOTE) {
+		if (!amount_msat_sub_sat(&receivable, receivable,
+					 commit_txfee(channel, receivable)))
+			receivable = AMOUNT_MSAT(0);
+	}
+	/* They can't offer an HTLC less than the we will accept. */
```
Comment (cdecker) [inline lightningd/peer_control.c]:
```suggestion
	/* They can't offer an HTLC less than what we will accept. */
```

File: doc/lightning-listpeers.7.md [hunk 8]
```diff
@@ -153,15 +153,15 @@ state, or in various circumstances:
   via this channel;
   a number followed by a string unit.
   The peer imposes this on us, default is 1% of the total channel capacity.
-* *spendable\_msat*: A string describing an ***estimate*** of how much we
-  can send out over this channel in a single payment (or payment-part for
-  multi-part payments);
+* *spendable\_msat* and *receivable\_msat*: A string describing an
+  ***estimate*** of how much we can send or receive over this channel in a
+  single payment (or payment-part for multi-part payments);
```
Comment (cdecker) [inline doc/lightning-listpeers.7.md]: Good catch on the "single payment" part :+1:

File: tests/test_pay.py [hunk 9]
```diff
@@ -2256,26 +2309,31 @@ def test_channel_spendable_large(node_factory, bitcoind):
     payment_hash = l2.rpc.invoice('any', 'inv', 'for testing')['payment_hash']
 
     # We should be able to spend this much, and not one msat more!
-    amount = l1.rpc.listpeers()['peers'][0]['channels'][0]['spendable_msat']
+    spendable = l1.rpc.listpeers()['peers'][0]['channels'][0]['spendable_msat']
+
+    # receivable from the other side should calculate to the exact same amount
+    receivable = l1.rpc.listpeers()['peers'][0]['channels'][0]['spendable_msat']
```
Comment (cdecker) [inline tests/test_pay.py]:
```suggestion
    receivable = l2.rpc.listpeers()['peers'][0]['channels'][0]['receivable_msat']
```

File: doc/STYLE.md [hunk 10]
```diff
@@ -36,6 +36,22 @@ using early returns or continues, eg:
...
+arguments like so:
+
+```C
+static void subtract_received_htlcs(const struct channel *channel,
+				    struct amount_msat *amount)
+```
+
+Note: For more details, the files `.clang-format` and `.editorconfig` are
+located in the projects root directory.
+
```
Comment (cdecker) [inline doc/STYLE.md]: This seems like an unrelated change to the rest of the PR. I usually collect these kinds of commits in a weekly cleanup PR :wink:

File: tests/test_pay.py (hunk 4 above)
Comment (cdecker) [reply tests/test_pay.py]: Internal multiplication followed by some rounding maybe?

Comment (m-schmoock) [reply tests/test_pay.py]: Considered 'resolved' by raising issue https://github.com/ElementsProject/lightning/issues/3595

File: doc/STYLE.md (hunk 10 above)
Comment (m-schmoock) [reply doc/STYLE.md]: Done by https://github.com/ElementsProject/lightning/pull/3598

File: tests/test_pay.py (hunk 9 above)
Comment (m-schmoock) [reply tests/test_pay.py]: thx. good catch

File: lightningd/peer_control.c (hunk 6 above)
Comment (m-schmoock) [reply lightningd/peer_control.c]: Okay, I left the old style, my reasoning was to try to make the two partly identical blocks visually distinguishable.

**Comment by m-schmoock** [comment]
> We might want to extract the LOCAL/REMOTE parameter into the function signature, and then internally use side and !side whenever either is used in the function. This is already what we do in a couple of places, and the enum is such that !LOCAL == REMOTE and vice-versa :-)

I'll check if I can come up with something...

**Comment by m-schmoock** [comment]
@cdecker Is there more I can do? I.e. merge the two `spendable` `receivable` blocks line 725 and following ( https://github.com/ElementsProject/lightning/blob/2b4f0341fdbf1cb8378c8f75765d29fd5074fd78/lightningd/peer_control.c#L725-L784 ) with all the duplicated comments like this:

```C
	/* Compute how much we can send/receive via this channel in one payment. */
	if (!amount_msat_sub_sat(&spendable, channel->our_msat,
				 channel->channel_info.their_config.channel_reserve))
		spendable = AMOUNT_MSAT(0);
	if (!amount_sat_sub_msat(&their_msat, channel->funding, channel->our_msat))
		their_msat = AMOUNT_MSAT(0);
	if (!amount_msat_sub_sat(&receivable, their_msat,
				 channel->our_config.channel_reserve))
		receivable = AMOUNT_MSAT(0);

	/* Take away any currently-offered HTLCs. */
	subtract_offered_htlcs(channel, &spendable);
	subtract_received_htlcs(channel, &receivable);

	/* Substract the funders txfees */
	if (channel->funder == LOCAL) {
		if (!amount_msat_sub_sat(&spendable, spendable,
					 commit_txfee(channel, spendable,
						      LOCAL)))
			spendable = AMOUNT_MSAT(0);
	} else {
		if (!amount_msat_sub_sat(&receivable, receivable,
					 commit_txfee(channel, receivable, REMOTE)))
			receivable = AMOUNT_MSAT(0);
	}

	/* Can't offer an HTLC less than the other side will accept. */
	if (amount_msat_less(spendable,
			     channel->channel_info.their_config.htlc_minimum))
		spendable = AMOUNT_MSAT(0);
	if (amount_msat_less(receivable, channel->our_config.htlc_minimum))
		receivable = AMOUNT_MSAT(0);

	/* Can't offer an HTLC over the max payment threshold either. */
	if (amount_msat_greater(spendable, chainparams->max_payment))
		spendable = chainparams->max_payment;
	if (amount_msat_greater(receivable, chainparams->max_payment))
		receivable = chainparams->max_payment;

	/* append spendable/receivable to JSON output */
	json_add_amount_msat_compat(response, spendable,
				    "spendable_msatoshi", "spendable_msat");
	json_add_amount_msat_compat(response, receivable,
				    "receivable_msatoshi", "receivable_msat");
```

**Comment by cdecker** [comment]
ACK https://github.com/ElementsProject/lightning/pull/3572/commits/2b4f0341fdbf1cb8378c8f75765d29fd5074fd78
